src/Codegen/Constraints/BaseBuilder.php

```diff
@@ -92,6 +93,66 @@ protected function addEnumConstraintCheck(HackBuilder $hb): void {
     }
   }
 
+  protected function addHackEnumConstraintCheck(HackBuilder $hb): void {
+    $schema = type_assert_type($this->typed_schema, TSchema::class);
+    if (!Shapes::keyExists($schema, 'hackEnum')) {
+      return;
+    }
+
+    try {
+      $rc = new \ReflectionClass($schema['hackEnum']);
+    } catch (\ReflectionException $e) {
+      throw new \Exception(Str\format("Hack enum '%s' does not exist", $schema['hackEnum']));
+    }
+
+    invariant($rc->isEnum(), "'%s' is not an enum", $schema['hackEnum']);
+
+    $schema_type = $schema['type'] ?? null;
+    $hack_enum_values = keyset[];
+    foreach ($rc->getConstants() as $hack_enum_value) {
+      if ($schema_type === TSchemaType::INTEGER_T) {
+        $hack_enum_value = $hack_enum_value ?as int;
+      } else {
+        $hack_enum_value = $hack_enum_value ?as string;
+      }
+      invariant(
+        $hack_enum_value is nonnull,
+        "'%s' must contain only values of type %s",
+        $rc->getName(),
+        $schema_type === TSchemaType::INTEGER_T ? 'int' : 'string'
+      );
+      $hack_enum_values[] = $hack_enum_value;
+    }
+
+    if (Shapes::keyExists($schema, 'enum')) {
+      // If both `enum` and `hackEnum` are specified, assert that `enum` is a subset of
+      // `hackEnum` values. Any value not also in `hackEnum` can't be valid.
+      foreach ($schema['enum'] as $enum_value) {
```

Comment:
Oh, I didn't realize we only run that test when schemas change. I think even knowing that, the worst that can happen is that `enum` contains values not in `hackEnum`, causing the validator to reject those values. While not great, that's probably better than accepting the values.

What do you think of removing this check altogether? Maybe the framework shouldn't care if `enum` and `hackEnum` are out of sync, especially since it can't guarantee that they are.

Reply:
Per a conversation on Slack: we agreed the validation as-is is better than no validation.